src/knode/upstream.clj

```diff
@@ -0,0 +1,126 @@
+(ns knode.upstream
+  (:require
+   [clojure.java.io :as io]
+   [clojure.data :as dat]
+
+   digest
+   [tempfile.core :as tmp]
+   [org.httpkit.client :as http]
+   [clojure.data.xml :as xml]))
+
+"http://purl.obolibrary.org/obo/chebi.owl"
+"ftp://ftp.ebi.ac.uk/pub/databases/chebi/ontology/chebi.owl"
+"http://purl.obolibrary.org/obo/mro.owl"
+"https://raw.githubusercontent.com/IEDB/MRO/v2016-12-15/mro.owl"
+
+(require '[knode.upstream :as up])
+(require '[clojure.data.xml :as xml])
+(def parsed (xml/parse (java.io.StringReader. (up/slurp-gzipped "tmp/obo/mro/2016-12-15/mro.owl"))))
+(do (def ont (slurp "tmp/obo/chebi/151/chebi-raw.owl")) nil)
+
+(def upstream-meta
+  (atom
+   (if (.exists (io/file "knode-meta.edn"))
+     (read-string (slurp "knode-meta.edn"))
+     {})))
+(defn store-upstream-meta! [iri meta]
+  (swap! upstream-meta #(assoc % iri meta))
+  (spit "knode-meta.edn" @upstream-meta)
+  nil)
+
+(defn xml->version-iri [stream]
+  (->> stream
+       xml/parse
+       :content first :content (filter #(= :versionIRI (:tag %)))
+       first :attrs :rdf/resource))
+
+(defn xml-string->version-iri [string]
+  (xml->version-iri (java.io.StringReader. string)))
+
+(defmulti spit-gzipped! #(class %2))
+
+(defmethod spit-gzipped! java.io.BufferedReader [path content]
+  (io/make-parents path)
+  (with-open [s (-> path
+                    clojure.java.io/output-stream
+                    java.util.zip.GZIPOutputStream.
+                    clojure.java.io/writer)]
+    (binding [*out* s]
+      (loop [ln (.readLine content)]
+        (when ln
+          (println ln)
+          (recur (.readLine content)))))))
+
+(defmethod spit-gzipped! java.lang.String [path content]
+  (spit-gzipped! path (java.io.BufferedReader. (java.io.StringReader. content))))
+
+(defn slurp-gzipped [path]
+  (with-open [in (java.util.zip.GZIPInputStream. (io/input-stream path))]
+    (slurp in)))
+
+(defn ->upstream-path [prefix iri]
+  (io/as-relative-path (str prefix (.getPath (java.net.URL. iri)))))
```

Comment:
I would like the compressed files to have the usual file extension, e.g. `mro.owl.gz`.